src/fault/Action.cpp

```diff
@@ -59,66 +66,185 @@ void Action::ensure(Type t)
         throw "wrong action type";
 }
 
-Action::Action() : type_(NOP)
+bool Action::is_action_on_field(void) const
+{
+    return ((type_ == Type::BITFLIP || type_ == Type::MASK) ? true : false);
+}
+
+Action::Action() : type_(Type::NOP)
 {
     etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action() called. "));
 }
 
-Action::Action(const InjectorAddress &inj, const std::string &command) : type_(COMMAND), inj_(inj), command_(command)
+#ifndef NO_ETISS
+Action::Action(int32_t event) : type_(Type::EVENT), event_(event)
+{
+    etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action(int32_t exception) called. "));
+}
+#endif
+
+Action::Action(const InjectorAddress &inj, const std::string &command)
+    : type_(Type::COMMAND), inj_(std::make_unique<InjectorAddress>(inj)), command_(command)
 {
     etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action(InjectorAddress &=") + inj.getInjectorPath() +
                                    std::string(", command=") + command + std::string(") called. "));
 }
 
 Action::Action(const InjectorAddress &inj, const std::string &field, unsigned bit)
-    : type_(BITFLIP), inj_(inj), field_(field), bit_(bit)
+    : type_(Type::BITFLIP), inj_(std::make_unique<InjectorAddress>(inj)), field_(field), bit_(bit)
 {
     etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action(InjectorAddress &=") + inj.getInjectorPath() +
                                    std::string(", field=") + field + std::string(", bit=") + std::to_string(bit) +
                                    std::string(") called. "));
 }
 
-Action::Action(const Fault &fault) : type_(INJECTION)
+Action::Action(const InjectorAddress &inj, const std::string &field, MaskOp mask_op, uint64_t mask_value)
+    : type_(Type::MASK)
+    , inj_(std::make_unique<InjectorAddress>(inj))
+    , field_(field)
+    , mask_op_(mask_op)
+    , mask_value_(mask_value)
+{
+    etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action(InjectorAddress &=") + inj.getInjectorPath() +
+                                   std::string(", field=") + field + std::string(", mask_op=") + std::string(mask_op_) +
+                                   std::string(", mask_value=") + std::to_string(mask_value) +
+                                   std::string(") called. "));
+}
+
+Action::Action(const FaultRef &fault_ref, type_t type) : type_(type), fault_ref_(std::make_unique<FaultRef>(fault_ref))
+{
+    etiss::log(etiss::VERBOSE, std::string("etiss::fault::Action::Action(FaultRef &=") + fault_ref.toString() +
+                                   std::string(") called. "));
+}
+
+Action::Action(const Action &cpy)
+{
+    *this = cpy;
+}
+
+Action &Action::operator=(const Action &cpy)
+{
+    type_ = cpy.type_;
+    switch (type_)
+    {
+    case Type::BITFLIP:
+        inj_ = std::make_unique<InjectorAddress>(cpy.getInjectorAddress());
```

Comment:
it is not clear to me why you made inj_ pointers. could they not just be value types so that this copy becomes a trivial "=default"?

Reply:
Unfortunately not so easily. 
For example in Stressor::firedTrigger:
```
        for (std::vector<etiss::fault::Action>::iterator iter = find->second.actions.begin();
             iter != find->second.actions.end(); ++iter)
        {
...
            default: // on field actions
            {
                if (iter->getInjectorAddress().getInjector())
```
The `inj_` member of `Action` is used to decide whether the `Action` may run through this Injector or is legal in the first place. The check is whether the member was set. Without major changes to most of the old fault injection stuff, I can not re-use most of it. All this FI code is way older than this open source repository. I do not even know who wrote it initially.

The idea behind this PR was to make the FI usable again, seems to take a lot more work to make the code up to date than the functionality.